### PR TITLE
PYTHON-4725, PYTHON-4860, PYTHON-4871, PYTHON-4843

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
Looks like we're stilling seeing the [same issue we saw here](https://github.com/mongodb/mongo-python-driver/pull/1896#issuecomment-2406249318):
```
 [2024/10/18 12:09:53.894] test/asynchronous/test_collection_management.py::TestCollectionManagementTimeseriesCollection::test_insertMany_with_duplicate_ids SKIPPED [  6%]
 [2024/10/18 12:34:53.894] Timeout (0:25:00)!
 [2024/10/18 12:34:53.894] Thread 0x00007f976d63b700 (most recent call first):
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/concurrent/futures/thread.py", line 89 in _worker
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 992 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1041 in _bootstrap_inner
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1012 in _bootstrap
 [2024/10/18 12:34:53.894] Thread 0x00007f976eb3b700 (most recent call first):
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/concurrent/futures/thread.py", line 89 in _worker
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 992 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1041 in _bootstrap_inner
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1012 in _bootstrap
 [2024/10/18 12:34:53.894] Thread 0x00007f977f304700 (most recent call first):
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/concurrent/futures/thread.py", line 89 in _worker
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 992 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1041 in _bootstrap_inner
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1012 in _bootstrap
 [2024/10/18 12:34:53.894] Thread 0x00007f977ea03700 (most recent call first):
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/concurrent/futures/thread.py", line 89 in _worker
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 992 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1041 in _bootstrap_inner
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1012 in _bootstrap
 [2024/10/18 12:34:53.894] Thread 0x00007f977fc05700 (most recent call first):
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/concurrent/futures/thread.py", line 89 in _worker
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 992 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1041 in _bootstrap_inner
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/threading.py", line 1012 in _bootstrap
 [2024/10/18 12:34:53.894] Thread 0x00007f978fb12740 (most recent call first):
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/traceback.py", line 315 in __init__
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/traceback.py", line 483 in _extract_from_extended_frame_gen
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/traceback.py", line 445 in extract
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/traceback.py", line 260 in extract_stack
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/base_events.py", line 460 in create_future
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/tasks.py", line 515 in _wait
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/tasks.py", line 451 in wait
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/network_layer.py", line 269 in async_receive_data
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/network.py", line 335 in receive_message
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/network.py", line 203 in command
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/pool.py", line 536 in command
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/helpers.py", line 45 in inner
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/pool.py", line 423 in _hello
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/monitor.py", line 363 in _check_with_socket
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/monitor.py", line 316 in _check_once
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/monitor.py", line 248 in _check_server
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/monitor.py", line 205 in _run
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/asynchronous/monitor.py", line 75 in target
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/pymongo/periodic_executor.py", line 101 in _run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/events.py", line 89 in _run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/base_events.py", line 2019 in _run_once
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/base_events.py", line 679 in run_forever
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/base_events.py", line 708 in run_until_complete
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/asyncio/runners.py", line 118 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/unittest/async_case.py", line 115 in _callMaybeAsync
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/unittest/async_case.py", line 93 in _callTestMethod
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/unittest/case.py", line 651 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/unittest/async_case.py", line 134 in run
 [2024/10/18 12:34:53.894]   File "/opt/python/3.13/lib/python3.13/unittest/case.py", line 707 in __call__
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/unittest.py", line 332 in runtest
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/runner.py", line 174 in pytest_runtest_call
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_callers.py", line 103 in _multicall
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_manager.py", line 120 in _hookexec
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_hooks.py", line 513 in __call__
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/runner.py", line 242 in <lambda>
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/runner.py", line 341 in from_call
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/runner.py", line 241 in call_and_report
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/runner.py", line 132 in runtestprotocol
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/runner.py", line 113 in pytest_runtest_protocol
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_callers.py", line 103 in _multicall
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_manager.py", line 120 in _hookexec
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_hooks.py", line 513 in __call__
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/main.py", line 362 in pytest_runtestloop
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_callers.py", line 103 in _multicall
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_manager.py", line 120 in _hookexec
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_hooks.py", line 513 in __call__
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/main.py", line 337 in _main
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/main.py", line 283 in wrap_session
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
 [2024/10/18 12:34:53.894]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_callers.py", line 103 in _multicall
 [2024/10/18 12:34:53.895]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_manager.py", line 120 in _hookexec
 [2024/10/18 12:34:53.895]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pluggy/_hooks.py", line 513 in __call__
 [2024/10/18 12:34:53.895]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/config/__init__.py", line 175 in main
 [2024/10/18 12:34:53.895]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/_pytest/config/__init__.py", line 201 in console_main
 [2024/10/18 12:34:53.895]   File "/data/mci/7ac5065bea1ac425a5a883530bb81a6a/src/.hatch/data/env/virtual/test/lib/python3.13/site-packages/pytest/__main__.py", line 9 in <module>
 [2024/10/18 12:34:53.895]   File "<frozen runpy>", line 88 in _run_code
 [2024/10/18 13:04:33.238] Hit exec timeout (1h0m0s).
 [2024/10/18 13:04:33.238] test/asynchronous/test_command_logging.py::TestCommandLoggingCommand::test_A_failed_command
 [2024/10/18 13:04:33.276]   File "<frozen runpy>", line 198 in _run_module_as_main
 [2024/10/18 13:04:33.296] Command 'shell.exec' in function 'run tests' (step 2.2 of 2) stopped early: context canceled.
```
https://spruce.mongodb.com/task/mongo_python_driver_test_rhel8_py3.13_auth_ssl_cov_test_4.2_sharded_cluster_patch_a1ade45dd3b67a6e4baa50404b9807f68062f043_6712b06f0873c20007c7d874_24_10_18_19_01_09/logs?execution=0&sortBy=STATUS&sortDir=ASC

**Comment by NoahStapp** [comment]
We're still seeing the same issue, investigating further.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Some of the encryption tests are failing on 4.0:
```
 [2024/11/20 13:36:49.569] FAILURE: pymongo.errors.ConfigurationError: Auto-encryption requires a minimum MongoDB version of 4.2 ()
 [2024/11/20 13:36:49.569] self = <test.test_encryption.TestDecryptProse testMethod=test_04_decrypt_success>
 [2024/11/20 13:36:49.569]     def test_04_decrypt_success(self):
 [2024/11/20 13:36:49.569] >       self.encrypted_client.db.decryption_events.insert_one({"encrypted": self.cipher_text})
 [2024/11/20 13:36:49.569] test/test_encryption.py:1878: 
```
https://spruce.mongodb.com/task/mongo_python_driver_encryption_rhel8_python3.10_test_4.0_sharded_cluster_auth_ssl_sync_async_patch_1c7a7fe9ec3119228bc7bf98f1f9de199a4f8f2c_673e5143859d330007f67f84_24_11_20_21_14_44?execution=0&sortBy=STATUS&sortDir=ASC

That sounds like an issue with the new _setup_class code.

File: test/asynchronous/test_encryption.py [hunk 1]
```diff
@@ -1234,7 +1234,9 @@ async def test_03_bulk_batch_split(self):
         doc2 = {"_id": "over_2mib_2", "unencrypted": "a" * _2_MiB}
         self.listener.reset()
         await self.coll_encrypted.bulk_write([InsertOne(doc1), InsertOne(doc2)])
-        self.assertEqual(self.listener.started_command_names(), ["insert", "insert"])
+        self.assertEqual(
+            len([c for c in self.listener.started_command_names() if c == "insert"]), 2
+        )
```
Comment (ShaneHarvey) [inline test/asynchronous/test_encryption.py:1239]: What other commands are interleaved here?

**Review by ShaneHarvey — APPROVED** [review]

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**